README.md
# Govee Temperature/Humidity BLE Home Assistant Component

**<u>NOTICE Jul, 15,2021</u>:**  
**Govee support has been added to [Passive BLE Monitor integration](https://github.com/custom-components/ble_monitor) as of version 4.1.0.**

For any new users I would strongly suggest using that integration.  BLE Monitor has better support Bluetooth support (including Intel NUC), more sensors and sensor discovery so there is no longer a need to list MAC addresses.  

Existing users: I will continue to support this component for the foreseeable future but new development will likely only occur for [Passive BLE Monitor integration](https://github.com/custom-components/ble_monitor).  Sadly, you will not be able to use both components at the same time.

Thank you @Ernst79 for adding the Govee devices.  Ultimately, it makes more sense for you, the end users, to have one component that does everything well and to pool development resources to ensure issues can be resolved quickly.

## Govee Temperature/Humidity BLE Home Assistant Component

A custom component for [Home Assistant](https://www.home-assistant.io) that listens for the advertisement message broadcast by Govee Bluetooth Thermometer/Hygrometers.

[![hacs_badge](https://img.shields.io/badge/HACS-Custom-orange.svg?style=plastic)](https://github.com/hacs/integration)
![hassfest_badge](https://github.com/Home-Is-Where-You-Hang-Your-Hack/sensor.goveetemp_bt_hci/actions/workflows/hassfest.yaml/badge.svg)

## Supported Devices
* Govee H5051 (BLE only)
* Govee H5072
* [Govee H5074](https://www.amazon.com/Govee-Thermometer-Hygrometer-Bluetooth-Temperature/dp/B07R586J37)
* [Govee H5075](https://www.amazon.com/Govee-Temperature-Humidity-Notification-Monitor/dp/B07Y36FWTT/)
* Govee H5101
* [Govee H5102](https://www.amazon.com/gp/product/B087313N8F/)
* [Govee H5177](https://www.amazon.com/gp/product/B08C9VYMHY)
* [Govee H5179](https://www.amazon.com/gp/product/B0872ZWV8X) (BLE only)

## Installation


**1. Install the custom component:**

- The easiest way is to install it with [HACS](https://hacs.xyz/). First install [HACS](https://hacs.xyz/) if you don't have it yet. After installation, the custom component can be found in the HACS store under integrations.

- Alternatively, you can install it manually. Just copy paste the content of the `sensor.goveetemp_bt_hci/custom_components` folder in your `config/custom_components` directory.
     As example, you will get the `sensor.py` file in the following path: `/config/custom_components/govee_ble_hci/sensor.py`.

*NOTE:* the following instructions about setting device permissions are an edge case for a very specific set up.  (If you do not understand it, do not worry about).

- If running Home Assistant without root access the [Bleson](https://github.com/TheCellule/python-bleson) Python library used for accessing bluetooth requires the following permissions applied to the Python 3 binary. If using a virtual environment for HA, this binary will be in the virtual environment path.

     *NOTE*: Replace "path" with the path to the Python3 binary (example: /srv/homeassistant/bin/python3)
     ```
     sudo setcap cap_net_raw,cap_net_admin+eip $(eval readlink -f path)
     ```
**2. Stop and start Home Assistant:**

- Stop and start Home Assistant. Make sure you first stop Home Assistant and then start Home Assistant again.  Do this before step 5, as Home Assistant will otherwise complain that your configuration is not valid (as it still uses the build in `govee_ble_hci` integration), and won't restart when hitting restart in the server management menu.

**3. Add the platform to your configuration.yaml file (see [below](#configuration))**

**4. Restart Home Assistant:**

- A second restart may be required to load the configuration. Within a few minutes, the sensors should be added to your home-assistant automatically (at least two [period](#period) may be required.  If the [period](#period) is set to a time greater than two minutes, at least four [period](#period) may be required).

**5. If the entities are still not displaying data, a restart of the host device may be required.**

### Troubleshooting and help

Any questions or support should be asked on [this component's Home Assistant community post](https://community.home-assistant.io/t/govee-ble-thermometer-hygrometer-sensor/166696).

### Configuration Variables

Specify the sensor platform `govee_ble_hci` and a list of devices with unique MAC address.

*NOTE*: device name is optional.  If not provided, devices will be labeled using the MAC address
```
sensor:
  - platform: govee_ble_hci
    govee_devices:
      - mac: "A4:C1:38:A1:A2:A3"
        name: Bedroom
      - mac: "A4:C1:38:B1:B2:B3"
      - mac: "A4:C1:38:C1:C2:C3"
        name: Kitchen
```



##### Additional component configuration options

| Option | Type |Default Value | Description |  
| -- | -- | -- | -- |
| `rounding`| Boolean | `True` | Enable/disable rounding of the average of all measurements taken within the number seconds specified with 'period'. |  
| `decimals` | positive integer | `2`| Number of decimal places to round if rounding is enabled. NOTE: the raw Celsius is rounded and setting `decimals: 0` will still result in decimal values returned for Fahrenheit as well as temperatures being off by up to 1 degree `F`.|
| `period` | positive integer | `60` | The period in seconds during which the sensor readings are collected and transmitted to Home Assistant after averaging. The Govee devices broadcast roughly once per second so this limits amount of mostly duplicate data stored in  Home Assistant's database. |
| `log_spikes` |  Boolean | `False` | Puts information about each erroneous spike in the Home Assistant log. |
| `use_median` | Boolean  | `False` | Use median as sensor output instead of mean (helps with "spiky" sensors). Please note that both the median and the mean values in any case are present as the sensor state attributes. |
| `hci_device`| string | `hci0` | HCI device name used for scanning. |
| `temp_range_min_celsius` | float | `-20.0` | Set the lower bound of reasonable measurements, in Celsius. Temperature measurements lower than this will be discarded. *Warning*: temperatures returned by the Govee device that are outside of the specified range may not be accurate.  It is not advised to change this value.|
| `temp_range_max_celsius` | float | `60.0` | Set the upper bound of reasonable measurements, in Celsius. Temperature measurements higher than this will be discarded. *Warning*: temperatures returned by the Govee device that are outside of the specified range may not be accurate.  It is not advised to change this value.|

Example with all defaults:
```
sensor:
  - platform: govee_ble_hci
    rounding: True
    decimals: 2
    period: 60
    log_spikes: False
    hci_device: hci0
    govee_devices:
      - mac: "A4:C1:38:A1:A2:A3"
        name: Bedroom
      - mac: "A4:C1:38:B1:B2:B3"
      - mac: "A4:C1:38:C1:C2:C3"
        name: Kitchen
```

## Credits
  This was originally based on/shamelessly copied from [custom-components/sensor.mitemp_bt](https://github.com/custom-components/sensor.mitemp_bt).  I want to thank [@tsymbaliuk](https://community.home-assistant.io/u/tsymbaliuk) and [@Magalex](https://community.home-assistant.io/u/Magalex) for providing a blueprint for developing my Home Assistant component.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 0.10.3
**Chore:**
  - **Remove deprecated `device_class` properties.**

## 0.10.2
**Fix:**
  - **Refactor to use SensorEntityDescription for future Home Assistant compatibility**

**Possible Breaking Change**
For those who display temperature fahrenheit, due a change with how Home Assistant handles sensors, your existing sensors may display in Celsius after this update.
 Under Configuration -> Devices & Services -> Entities tab,  the settings for each temperature sensor can be modified to display F as the unit of measurement.

## 0.10.1
**Fix:**
  - **Renamed `device_state_attributes` to `extra_state_attributes`.  Remove warnings related to Home Assistant [PR#47304](https://github.com/home-assistant/core/pull/47304)**

## 0.10.0
**Feat:**
  - **Minimum and maximum temperature configurable.**  Base on [SteveOnorato/moat_temp_hum_ble](https://github.com/SteveOnorato/moat_temp_hum_ble/).  Thank you @SteveOnorato
  - **Devices now related.** Thank you @natekspencer
**Fix:**
  - **Catch Bluetooth adapter error and provide useful information.**
**Chore:**
  - **Removed "CONF_HCITOOL_ACTIVE"**

**BREAKING CHANGE** - With the removal of `CONF_HCITOOL_ACTIVE` if `hcitool_active` is still in your config, Home Assistant will not start.  Please remove this deprecated flag.

## 0.9.2
**Fix:**
  - **Update bleson package to version 0.18, fixes non BLE spec device name errors**

## 0.9.1
**Fix:**
  - **Add version to manifest.json**

## 0.9.0
**Feature:**
  - **Added support for the Govee H5179** (Thank you [skilau](https://github.com/skilau))

## 0.8.0
**Fix:**
  - **Decode negative temperature for Govee H5072, Govee H5075, Govee H5051 and Govee H5052**
  - **Catch remaining StatisticsError instances in sensor**

## 0.7.1
**Fix:**
  - **Update bleson package, fixes constant RSSI value**

## 0.7
**Feature:**
  - **Added support for the Govee H5051** (Thank you [billprozac](https://github.com/billprozac))
  - **Added support for the Govee H5102** (Thank you [billprozac](https://github.com/billprozac))

**Fix:**
  - **Restart scanning each period to prevent device sleeping** (Thank you [billprozac](https://github.com/billprozac))

**Docs:**
  - **Added non-root user note** (Thank you [spinningmonkey](https://github.com/spinningmonkey))

## 0.6
**Fix:**
  - **Removed hcitool dependencies.**
  - **Restructure component for easier maintenance.**
  - **Deprecated `hcitool_active` configuration option.**

## 0.5
**Feature:**
  - **Added support for the Govee H5072**

## 0.4
**Fixed:**

 - **Correct two's complement conversion**
 - **Initialize m_temp and m_hum to prevent UnboundLocalError error**

## 0.3
**Fixed:**

 - **Prevent crashing on reading of corrupt broadcast**

## 0.2

**Fixed:**

 - **Fix limit updates to period interval** (Thank you [sfjes](github.com/sfjes))

## 0.1
  - **Initial Release**


custom_components/govee_ble_hci/ble_ht.py
"""Bluetooth LE Humidity/Temperature data classes."""
from typing import List, Optional, Union
import statistics as sts
import logging

from .const import (
    DEFAULT_TEMP_RANGE_MIN,
    DEFAULT_TEMP_RANGE_MAX,
    CONF_HMIN,
    CONF_HMAX,
)

_LOGGER = logging.getLogger(__name__)


class BLE_HT_packet:
    """Bluetooth LE Humidity/Temperature packet data."""

    temperature: float
    humidity: float
    packet: str


class BLE_HT_data:
    """Bluetooth LE Humidity/Temperature data."""

    _desc: Optional[str]
    _mac: str
    _rssi: List[int]
    _battery: Optional[int]
    _packet_data: List[BLE_HT_packet]
    _decimal_places: Optional[int]
    _log_spikes: bool
    _min_temp: float
    _max_temp: float
    _sensor_number: Optional[int]

    def __init__(
        self,
        mac: str,
        description: Optional[str],
        sensor_number: Optional[int]
    ) -> None:
        """Init."""
        self._mac = mac
        self._desc = description
        self._log_spikes = False
        self._min_temp = DEFAULT_TEMP_RANGE_MIN
        self._max_temp = DEFAULT_TEMP_RANGE_MAX
        self._sensor_number = sensor_number
        self.reset()

    @property
    def data_size(self) -> int:
        """Packet data length."""
        return len(self._packet_data)

    @property
    def last_packet(self) -> Optional[str]:
        """Return MAC address."""
        if len(self._packet_data) == 0:
            return None
        return self._packet_data[-1].packet

    @property
    def mac(self) -> str:
        """Return MAC address."""
        return self._mac

    @property
    def battery(self) -> Optional[int]:
        """Return battery remaining value."""
        return self._battery

    @battery.setter
    def battery(self, value: Optional[int]) -> None:
        """Set battery remaining value."""
        if isinstance(value, int):
            self._battery = value

    @property
    def decimal_places(self) -> Optional[int]:
        """Set number of decimal places for rounding value."""
        return self._decimal_places

    @decimal_places.setter
    def decimal_places(self, value: int) -> None:
        """Set number of decimal places for rounding value."""
        if value >= 0:
            self._decimal_places = value

    @property
    def description(self) -> Optional[str]:
        """Return device description or MAC address."""
        return self._desc if hasattr(self, "_desc") else self._mac

    @description.setter
    def description(self, value: str) -> None:
        """Set device description."""
        if isinstance(value, str):
            self._desc = value

    @property
    def log_spikes(self) -> bool:
        """Set number of decimal places for rounding value."""
        return self._log_spikes

    @log_spikes.setter
    def log_spikes(self, value: bool) -> None:
        """Set number of decimal places for rounding value."""
        self._log_spikes = value

    @property
    def rssi(self) -> Optional[int]:
        """Return RSSI value."""
        try:
            return round(sts.mean(self._rssi))
        except (AssertionError, sts.StatisticsError):
            return None

    @rssi.setter
    def rssi(self, value: Optional[int]) -> None:
        """Set RSSI value."""
        if isinstance(value, int) and value < 0:
            self._rssi.append(value)

    @property
    def maximum_temperature(self) -> float:
        """Get upper bound of temperature."""
        return self._max_temp

    @maximum_temperature.setter
    def maximum_temperature(self, value: float) -> None:
        """Set upper bound of temperature."""
        self._max_temp = value

    @property
    def minimum_temperature(self) -> float:
        """Get lower bound of temperature."""
        return self._min_temp

    @minimum_temperature.setter
    def minimum_temperature(self, value: float) -> None:
        """Set lower bound of temperature."""
        self._min_temp = value

    @property
    def mean_temperature(self) -> Union[float, None]:
        """Mean temperature of values collected."""
        try:
            avg = sts.mean(self._map_packet_data_attrs("temperature"))
            if hasattr(self, "_decimal_places"):
                return round(avg, self._decimal_places)
            return avg
        except (AssertionError, sts.StatisticsError):
            return None

    @property
    def median_temperature(self) -> Union[float, None]:
        """Median temperature of values collected."""
        try:
            avg = sts.median(self._map_packet_data_attrs("temperature"))
            if hasattr(self, "_decimal_places"):
                return round(avg, self._decimal_places)
            return avg
        except (AssertionError, sts.StatisticsError):
            return None

    @property
    def mean_humidity(self) -> Union[float, None]:
        """Mean humidity of values collected."""
        try:
            avg = sts.mean(self._map_packet_data_attrs("humidity"))
            if hasattr(self, "_decimal_places"):
                return round(avg, self._decimal_places)
            return avg
        except (AssertionError, sts.StatisticsError):
            return None

    @property
    def median_humidity(self) -> Union[float, None]:
        """Median humidity of values collected."""
        try:
            avg = sts.median(self._map_packet_data_attrs("humidity"))
            if hasattr(self, "_decimal_places"):
                return round(avg, self._decimal_places)
            return avg
        except (AssertionError, sts.StatisticsError):
            return None

    @property
    def sensor_number(self) -> int:
        return self._sensor_number

    def update(
        self,
        temperature: Optional[float],
        humidity: Optional[float],
        packet: Optional[Union[int, str]],
    ) -> None:
        """Update packet data."""
        new_packet = BLE_HT_packet()

        # Check if temperature within bounds
        if temperature is not None and self._max_temp >= temperature >= self._min_temp:
            new_packet.temperature = temperature
        elif self._log_spikes:
            err = "Temperature spike: {} ({})".format(temperature, self._mac)
            _LOGGER.error(err)

        # Check if humidity within bounds
        if humidity is not None and CONF_HMAX >= humidity >= CONF_HMIN:
            new_packet.humidity = humidity
        elif self._log_spikes:
            err = "Humidity spike: {} ({})".format(humidity, self._mac)
            _LOGGER.error(err)

        new_packet.packet = str(packet)
        self._packet_data.append(new_packet)

    def reset(self) -> None:
        """Reset default values."""
        self._battery = None
        self._rssi = []
        self._packet_data = []

    def _map_packet_data_attrs(self, attr: str) -> List[float]:
        """Map defined values from _packet.data."""
        mapped_vals = []
        for datum in self._packet_data:
            if hasattr(datum, attr):
                mapped_vals.append(float(getattr(datum, attr)))
        return mapped_vals


custom_components/govee_ble_hci/sensor.py
"""Govee BLE monitor integration."""
from datetime import timedelta
import logging
import voluptuous as vol
from typing import Any, Collection, List, Optional, Dict, Sequence, Set, Tuple

from bleson import get_provider  # type: ignore
from bleson.core.hci.constants import EVT_LE_ADVERTISING_REPORT  # type: ignore
from bleson.core.types import BDAddress  # type: ignore
from bleson.core.hci.type_converters import hex_string  # type: ignore

from homeassistant.exceptions import HomeAssistantError  # type: ignore
from homeassistant.components.sensor import PLATFORM_SCHEMA  # type: ignore
import homeassistant.helpers.config_validation as cv  # type: ignore
from homeassistant.helpers.restore_state import RestoreEntity
from homeassistant.components.sensor import SensorEntity
from homeassistant.helpers.event import track_point_in_utc_time  # type: ignore
import homeassistant.util.dt as dt_util  # type: ignore
from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntityDescription,
    SensorStateClass,
)

from homeassistant.const import (  # type: ignore
    DEVICE_CLASS_TEMPERATURE,
    DEVICE_CLASS_HUMIDITY,
    TEMP_CELSIUS,
    PERCENTAGE,
    ATTR_BATTERY_LEVEL,
)

from .const import (
    CONF_DECIMALS,
    CONF_DEVICE_MAC,
    CONF_DEVICE_NAME,
    CONF_GOVEE_DEVICES,
    CONF_HCI_DEVICE,
    CONF_LOG_SPIKES,
    CONF_PERIOD,
    CONF_ROUNDING,
    CONF_TEMP_RANGE_MAX_CELSIUS,
    CONF_TEMP_RANGE_MIN_CELSIUS,
    CONF_USE_MEDIAN,
    DEFAULT_DECIMALS,
    DEFAULT_HCI_DEVICE,
    DEFAULT_LOG_SPIKES,
    DEFAULT_PERIOD,
    DEFAULT_ROUNDING,
    DEFAULT_TEMP_RANGE_MAX,
    DEFAULT_TEMP_RANGE_MIN,
    DEFAULT_USE_MEDIAN,
    DOMAIN,
)

from .govee_advertisement import GoveeAdvertisement
from .ble_ht import BLE_HT_data

###############################################################################

_LOGGER = logging.getLogger(__name__)

DEVICES_SCHEMA = vol.Schema(
    {
        vol.Optional(CONF_DEVICE_MAC): cv.string,
        vol.Optional(CONF_DEVICE_NAME): cv.string,
    }
)

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Optional(CONF_ROUNDING, default=DEFAULT_ROUNDING): cv.boolean,
        vol.Optional(CONF_DECIMALS, default=DEFAULT_DECIMALS): cv.positive_int,
        vol.Optional(CONF_PERIOD, default=DEFAULT_PERIOD): cv.positive_int,
        vol.Optional(CONF_LOG_SPIKES, default=DEFAULT_LOG_SPIKES): cv.boolean,
        vol.Optional(CONF_USE_MEDIAN, default=DEFAULT_USE_MEDIAN): cv.boolean,
        vol.Optional(CONF_GOVEE_DEVICES): vol.All([DEVICES_SCHEMA]),
        vol.Optional(CONF_HCI_DEVICE, default=DEFAULT_HCI_DEVICE): cv.string,
        vol.Optional(
            CONF_TEMP_RANGE_MIN_CELSIUS, default=DEFAULT_TEMP_RANGE_MIN
        ): float,
        vol.Optional(
            CONF_TEMP_RANGE_MAX_CELSIUS, default=DEFAULT_TEMP_RANGE_MAX
        ): float,
    }
)

###############################################################################

#
# Configure for Home Assistant
#
def setup_platform(hass, config, add_entities, discovery_info=None) -> None:
    """Set up the sensor platform."""
    _LOGGER.debug("Starting Govee HCI Sensor")

    govee_devices: List[BLE_HT_data] = []  # Data objects of configured devices
    sensors_by_mac = {}  # HomeAssistant sensors by MAC address
    adapter = None

    def handle_meta_event(hci_packet) -> None:
        """Handle recieved BLE data."""
        # If recieved BLE packet is of type ADVERTISING_REPORT
        if hci_packet.subevent_code == EVT_LE_ADVERTISING_REPORT:
            packet_mac = hci_packet.data[3:9]

            for device in govee_devices:
                # If recieved device data matches a configured govee device
                if BDAddress(device.mac) == BDAddress(packet_mac):
                    _LOGGER.debug(
                        "Received packet data for {}: {}".format(
                            BDAddress(device.mac), hex_string(hci_packet.data)
                        )
                    )
                    # parse packet data
                    ga = GoveeAdvertisement(hci_packet.data)

                    # If mfg data information is defined, update values
                    if ga.packet is not None:
                        device.update(ga.temperature, ga.humidity, ga.packet)

                    # Update RSSI and battery level
                    device.rssi = ga.rssi
                    device.battery = ga.battery

    def init_configureed_devices() -> None:
        """Initialize configured Govee devices."""
        for conf_dev in config[CONF_GOVEE_DEVICES]:
            # Initialize BLE HT data objects
            mac: str = conf_dev["mac"]
            given_name = conf_dev.get("name", None)

            device = BLE_HT_data(mac, given_name)
            device.log_spikes = config[CONF_LOG_SPIKES]
            device.maximum_temperature = config[CONF_TEMP_RANGE_MAX_CELSIUS]
            device.minimum_temperature = config[CONF_TEMP_RANGE_MIN_CELSIUS]

            if config[CONF_ROUNDING]:
                device.decimal_places = config[CONF_DECIMALS]
            govee_devices.append(device)

            # Initialize HA sensors
            name = conf_dev.get("name", mac)

            tempDescription = SensorEntityDescription(
                key="temperature",
                name=name,
                native_unit_of_measurement=TEMP_CELSIUS,
                device_class=SensorDeviceClass.TEMPERATURE,
                state_class=SensorStateClass.MEASUREMENT,
            )
  
            humDescription= SensorEntityDescription(
                key="humidity",
                name=name,
                native_unit_of_measurement=PERCENTAGE,
                device_class=SensorDeviceClass.HUMIDITY,
                state_class=SensorStateClass.MEASUREMENT,
            )
            temp_sensor = TemperatureSensor(mac, name,  tempDescription)
            hum_sensor = HumiditySensor(mac, name,  humDescription)
            sensors = [temp_sensor, hum_sensor]
            sensors_by_mac[mac] = sensors
            add_entities(sensors)

    def update_ble_devices(config) -> None:
        """Discover Bluetooth LE devices."""
        # _LOGGER.debug("Discovering Bluetooth LE devices")
        use_median = config[CONF_USE_MEDIAN]

        ATTR = "_device_state_attributes"
        textattr = "last median of" if use_median else "last mean of"

        for device in govee_devices:
            sensors = sensors_by_mac[device.mac]

            _LOGGER.debug(
                "Last mfg data for {}: {}".format(
                    BDAddress(device.mac), device.last_packet
                )
            )

            if device.last_packet:
                if device.median_humidity is not None:
                    humstate_med = float(device.median_humidity)
                    getattr(sensors[1], ATTR)["median"] = humstate_med
                    if use_median:
                        setattr(sensors[1], "_state", humstate_med)

                if device.mean_humidity is not None:
                    humstate_mean = float(device.mean_humidity)
                    getattr(sensors[1], ATTR)["mean"] = humstate_mean
                    if not use_median:
                        setattr(sensors[1], "_state", humstate_mean)

                if device.median_temperature is not None:
                    tempstate_med = float(device.median_temperature)
                    getattr(sensors[0], ATTR)["median"] = tempstate_med
                    if use_median:
                        setattr(sensors[0], "_state", tempstate_med)

                if device.mean_temperature is not None:
                    tempstate_mean = float(device.mean_temperature)
                    getattr(sensors[0], ATTR)["mean"] = tempstate_mean
                    if not use_median:
                        setattr(sensors[0], "_state", tempstate_mean)

                for sensor in sensors:
                    last_packet = device.last_packet
                    getattr(sensor, ATTR)["last packet id"] = last_packet
                    getattr(sensor, ATTR)["rssi"] = device.rssi
                    getattr(sensor, ATTR)[ATTR_BATTERY_LEVEL] = device.battery
                    getattr(sensor, ATTR)[textattr] = device.data_size
                    sensor.async_schedule_update_ha_state()

                device.reset()

    def update_ble_loop(now) -> None:
        """Lookup Bluetooth LE devices and update status."""
        _LOGGER.debug("update_ble_loop called")
        adapter.start_scanning()

        try:
            # Time to make the dounuts
            update_ble_devices(config)
        except RuntimeError as error:
            _LOGGER.error("Error during Bluetooth LE scan: %s", error)

        time_offset = dt_util.utcnow() + timedelta(seconds=config[CONF_PERIOD])
        # update_ble_loop() will be called again after time_offset
        track_point_in_utc_time(hass, update_ble_loop, time_offset)

    ###########################################################################

    # Initalize bluetooth adapter and begin scanning
    # XXX: will not work if there are more than 10 HCI devices
    try:
        adapter = get_provider().get_adapter(int(config[CONF_HCI_DEVICE][-1]))
        adapter._handle_meta_event = handle_meta_event
        hass.bus.listen("homeassistant_stop", adapter.stop_scanning)
        adapter.start_scanning()
    except (RuntimeError, OSError, PermissionError) as error:
        error_msg = "Error connecting to Bluetooth adapter: {}\n\n".format(error)
        error_msg += "Bluetooth adapter troubleshooting:\n"
        error_msg += "  -If running HASS, ensure the correct HCI device is being"
        error_msg += " used. Check by logging into HA command line and execute:\n"
        error_msg += "          gdbus introspect --system --dest org.bluez --object-path /org/bluez | fgrep -i hci\n"
        error_msg += "  -If running Home Assistant in Docker, "
        error_msg += "make sure it run with the --privileged flag.\n"
        # _LOGGER.error(error_msg)
        raise HomeAssistantError(error_msg) from error

    # Initialize configured Govee devices
    init_configureed_devices()
    # Begin sensor update loop
    update_ble_loop(dt_util.utcnow())


###############################################################################

#
# HomeAssistant Temperature Sensor Class
#
class TemperatureSensor(RestoreEntity, SensorEntity):
    """Representation of a sensor."""

    def __init__(self, mac: str, name: str, description: SensorEntityDescription):
        """Initialize the sensor."""
        self._state = None
        self._battery = None
        self._unique_id = "t_" + mac.replace(":", "")
        self._name = name
        self._mac = mac.replace(":", "")
        self._device_state_attributes = {}
        self.entity_description = description

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return "{} temp".format(self._name)

    @property
    def native_value(self):
        """Return the state of the sensor."""
        return self._state

    @property
    def native_unit_of_measurement(self) -> str:
        """Return the unit of measurement."""
        return TEMP_CELSIUS

    @property
    def device_info(self) -> Optional[Dict[str, Collection[Sequence[str]]]]:
        """Temperature Device Info."""
        return {
            "identifiers": {(DOMAIN, self._mac)},
            "name": self._name,
            "manufacturer": "Govee",
        }

    @property
    def should_poll(self) -> bool:
        """No polling needed."""
        return False

    @property
    def extra_state_attributes(self):
        """Return the state attributes."""
        return self._device_state_attributes

    @property
    def unique_id(self) -> str:
        """Return a unique ID."""
        return self._unique_id

    @property
    def force_update(self) -> bool:
        """Force update."""
        return True


#
# HomeAssistant Humidity Sensor Class
#
class HumiditySensor(RestoreEntity, SensorEntity):
    """Representation of a Sensor."""

    def __init__(self, mac: str, name: str, description: SensorEntityDescription):
        """Initialize the sensor."""
        self._state = None
        self._battery = None
        self._name = name
        self._unique_id = "h_" + mac.replace(":", "")
        self._mac = mac.replace(":", "")
        self._device_state_attributes = {}
        self.entity_description = description

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return "{} humidity".format(self._name)

    @property
    def state(self):
        """Return the state of the sensor."""
        return self._state

    @property
    def unit_of_measurement(self) -> str:
        """Return the unit of measurement."""
        return "%"

    @property
    def device_info(self) ->  Optional[Dict[str, Collection[Sequence[str]]]]:
        """Humidity Device Info."""
        return {
            "identifiers": {(DOMAIN, self._mac)},
            "name": self._name,
            "manufacturer": "Govee",
        }

    @property
    def should_poll(self) -> bool:
        """No polling needed."""
        return False

    @property
    def extra_state_attributes(self):
        """Return the state attributes."""
        return self._device_state_attributes

    @property
    def unique_id(self) -> str:
        """Return a unique ID."""
        return self._unique_id

    @property
    def force_update(self) -> bool:
        """Force update."""
        return True


custom_components/govee_ble_hci/const.py
"""Constants for the Govee BLE HCI monitor sensor integration."""

DOMAIN = "govee_ble_hci"

# Configuration options
CONF_DECIMALS = "decimals"
CONF_DEVICE_MAC = "mac"
CONF_DEVICE_NAME = "name"
CONF_DEVICE_SENSOR_NUMBER = "sensor_number"
CONF_GOVEE_DEVICES = "govee_devices"
CONF_HCI_DEVICE = "hci_device"
CONF_LOG_SPIKES = "log_spikes"
CONF_PERIOD = "period"
CONF_ROUNDING = "rounding"
CONF_TEMP_RANGE_MAX_CELSIUS = "temp_range_max_celsius"
CONF_TEMP_RANGE_MIN_CELSIUS = "temp_range_min_celsius"
CONF_USE_MEDIAN = "use_median"


# Default values for configuration options
DEFAULT_DECIMALS = 2
DEFAULT_HCI_DEVICE = "hci0"
DEFAULT_LOG_SPIKES = False
DEFAULT_PERIOD = 60
DEFAULT_ROUNDING = True
DEFAULT_TEMP_RANGE_MAX = 60.0
DEFAULT_TEMP_RANGE_MIN = -20.0
DEFAULT_USE_MEDIAN = False

"""Fixed constants."""

# Sensor measurement limits to exclude erroneous spikes from the results
CONF_HMIN = 0.0
CONF_HMAX = 99.9
